[알린] Vending step5

jinios opened the pull request:
옵저버 패턴 적용해보았습니다. 다만...MVC구조가 서로 독립되게 구현은 하지 못한것같습니다ㅠㅠ
1. 현재 구조는 금액이나 음료객체를 추가할때 뷰 컨트롤러에서 직접 "얼마" 혹은 "어떤 음료"까지 모델에 전달하고 모델은 이를 인자로 받아서 추가하는 동작을 하고있습니다. 
2. 그리고 추가 동작이 일어나면 그대로 view를 업데이트 하는 기능을 따로 하고있는데, 이때 notification을 사용합니다.

위의 사항중에서 1번처럼 동작하는 구조가 컨트롤러와 모델의 관계가 독립적이지 못한 것 같은데, 구조적으로 어떻게 접근하면 좋을까요? (제 생각엔 notification을 보내고 받을때 인자로 필요한 값을 전달하면 될 것 같긴한데..)
힌트가 필요합니다ㅠㅠㅠ...

jinios committed:
Update README - issue of singleton

jinios committed:
Implement Observer pattern
-Add custom extension: Notification.Name
-Add notification post in Vendingmachine class
-Add notification selector and observer in ViewController class

jinios committed:
Update README - step5

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -25,12 +25,21 @@ class ViewController: UIViewController {
         vending = VendingMachine.sharedVendingMachine()
```
vending 참조 변수를 그래도 사용하고 있네요. 오류는 이제 없나요? 
이제 vending 변수가 VendingMachine 타입에 의존하지 않도록 DIP를 적용하도록 개선해보세요. 
특히 여기처럼 VendingMachine.sharedVendingMachine()를 직접 참조하지 않도록 해야합니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -25,12 +25,21 @@ class ViewController: UIViewController {
 
     @IBAction func addButtonTouched(sender: UIButton) {
         guard let item = try? Controller.AdminController().addItem(sender.tag) else { return }
         vending.add(inputItem: item)
```
vending.add()를 하려고 AdminController에서 적합한 아이템을 가져오는거라면 addItem() 이름은 적합하지 않은 것 같네요. 
여기서 화면에 어떤 버튼이 어떤 item인지 찾아서 모델에 넘겨주는 동작이 어색하지는 않습니다.

godrm reviewed (commented):
수고하셨습니다. 다음 피드백을 확인해보세요.
View -> VC -> Model로 이벤트를 받아서 데이터를 업데이트하는 흐름과
Model -> VC -> View로 데이터가 바뀌고 화면이 갱신되는 흐름을 정리해보세요.

jinios committed:
Add SharingVendingMachine protocol

jinios committed:
Update README - observer pattern

jinios committed:
Rename function - addItem() to pickItem()

jinios commented:
피드백주신사항 수정했습니다. 
- 프로토콜을 만들어서, 뷰 컨트롤러에서 sharedInstance를 접근하는 방식을 self.sharedInstance로 접근했는데, 다만 이게 리뷰주신사항대로 프로토콜을 잘 활용한건지 잘 모르겠습니다..ㅠㅠ 이렇게 되면 AppDelegate에서 뷰컨트롤러에 자판기를 넘겨주는동작이 필요없긴합니다....
- addItem메소드는 pickItem으로 이름 변경했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -22,15 +27,24 @@ class ViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
-        vending = VendingMachine.sharedVendingMachine()
+        vending = self.sharedInstance()
```
오웃.. ViewController가 프로토콜을 채택할 필요까지는 없습니다 :(
var vending 변수가 VendingMachine 타입대신 사용할 인터페이스(메소드들)을 가지도록 프로토콜을 타입으로 쓰면 됩니다. 레벨2에서 User / Admin 모드를 위해서 프로토콜을 지정해놓지 않았던가요?

jinios committed:
Split AdminController from Controller struct

jinios committed:
Change type of vending property
- Add DefaultMode, AdminMode, UserMode protocols in VendingMachine
- Change type of variable vending in Viewcontroller into DefaultMode protocol type

jinios commented:
피드백 주신 사항 수정했습니다. 
- vendingMachine클래스 UserMode, AdminMode, DefaultMode 프로토콜 준수하도록 수정
- Appdelegate에서 vending 변수 넘겨줌!
- ViewController에 vending변수 타입 프로토콜로 변경
- viewDidLoad 외 viewcontroller 메소드에 옵셔널 처리 코드 추가

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachine.swift:
```diff
@@ -8,19 +8,48 @@
+    func add(inputItem: Beverage)
+    func removeItem(itemCode: Int) throws
+    func addBalance(money: Int)
+}
```
여기 AdminMode 나 UserMode 에 있는 메소드는 DefaultMode 프로토콜과 중복인데 일부러 미리 만들어 놓은건가요?

jinios replied:
네네 일단지금은 뷰 컨트롤러 하나에서 재고 추가랑 금액 확인 같은 동작을 하고있기때문에 DefaultMode에 다 넣어놨고, 뷰 컨트롤러가 분리돼서 프로토콜 컴포지션으로 타입 정해줄때 DefaultMode에서 중복되는 메소드들은 분리할예정입니다.

godrm commented:
오케이 다음 단계에서 개선해보세요.
